main/lsp/LSPTypechecker.cc

```diff
@@ -221,60 +306,9 @@ vector<core::FileRef> LSPTypechecker::runFastPath(LSPFileUpdates &updates, Worke
-            // Only set oldFoundHashesForFiles if symbols actually changed
-            // Means that no-op edits (and thus calls to LSPTypechecker::retypecheck) don't blow away
-            // methods only to redefine them with different IDs.
-
-            // Okay to `move` here (steals component of getFileHash) because we're about to use
-            // replaceFile to clobber fref.data(gs) anyways.
-            oldFoundHashesForFiles.emplace(fref, move(fref.data(*gs).getFileHash()->foundHashes));
-        }
-
-        gs->replaceFile(fref, updates.updatedFiles[idx]);
-        // If file doesn't have a typed: sigil, then we need to ensure it's typechecked using typed: false.
-        fref.data(*gs).strictLevel = pipeline::decideStrictLevel(*gs, fref, config->opts);
-
-        toTypecheck.emplace_back(fref);
-    }
-    fast_sort(toTypecheck);
-
-    config->logger->debug("Running fast path over num_files={}", toTypecheck.size());
-    unique_ptr<ShowOperation> op;
-    if (toTypecheck.size() > config->opts.lspMaxFilesOnFastPath / 2) {
-        op = make_unique<ShowOperation>(*config, ShowOperation::Kind::FastPath);
-    }
-    ENFORCE(gs->errorQueue->isEmpty());
-    vector<ast::ParsedFile> updatedIndexed;
-    for (auto &f : toTypecheck) {
-        // TODO(jvilk): We don't need to re-index files that didn't change.
-        // (`updates` has already-indexed trees, but they've been indexed with initialGS, not the
-        // `*gs` that we'll be typechecking with. We could do an ast::Substitute here if we had
-        // access to `initialGS`, but that's owned by the indexer thread, not this thread.)
-        auto t = pipeline::indexOne(config->opts, *gs, f);
-        updatedIndexed.emplace_back(ast::ParsedFile{t.tree.deepCopy(), t.file});
-        updates.updatedFinalGSFileIndexes.push_back(move(t));
-
-        // See earlier in the method for an explanation of the .empty() check here.
-        if (config->opts.lspExperimentalFastPathEnabled && !result.changedSymbolNameHashes.empty() &&
-            oldFoundHashesForFiles.find(f) == oldFoundHashesForFiles.end()) {
-            // This is an extra file that we need to typecheck which was not part of the original
-            // edited files, so whatever it happens to have in foundMethodHashes is still "old"
-            // (but we can't use `move` to steal it like before, because we're not replacing the
-            // whole file).
-            oldFoundHashesForFiles.emplace(f, f.data(*gs).getFileHash()->foundHashes);
-        }
-    }
-
-    ENFORCE(gs->lspQuery.isEmpty());
-    auto resolved =
-        config->opts.lspExperimentalFastPathEnabled
-            ? pipeline::incrementalResolve(*gs, move(updatedIndexed), std::move(oldFoundHashesForFiles), config->opts)
-            : pipeline::incrementalResolve(*gs, move(updatedIndexed), nullopt, config->opts);
+    auto [toTypecheck, oldFoundHashesForFiles] = filesToTypecheck(*gs, config, updates, workers, result);
```

Comment:
What's the reasoning for having `filesToTypecheck` be a separate function? If I'm reading the code correctly, it can be inlined into `runIncrementalResolver` below. This has a couple benefits:

- It means that we don't need two separate `runIncrementalResolver` methods (not sure if this was intentional or not, but this PR defines `runIncrementalResolver` as an overloaded function, which usually makes the code hard to follow when picking it back up 6 months in the future.)
- It means that the logic for getting the big list of old hashes can be in one method. Right now it's broken up, and I think that it's nice to be able to read one function top-to-bottom and see all the logic that it does to compute hashes. In particular there's some subtlety around whether we move or don't move the old found hashes. It would be nice to get it back to having that logic and comments being in one function.

Reply:
You're right here. I believe `filesToTypecheck` is a left over from one of my previous attempts to make incremental slow path work. Thanks for noticing this, I'll inline it